// testcafe-browser-provider-browserstack/lib/api-request.js
'use strict';

exports.__esModule = true;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

exports.default = function (apiPath, params) {
    if (!process.env['BROWSERSTACK_USERNAME'] || !process.env['BROWSERSTACK_ACCESS_KEY']) throw new Error(ERROR_MESSAGES.BROWSERSTACK_AUTHENTICATION_FAILED());

    var url = apiPath.url;

    var opts = {
        auth: {
            user: process.env['BROWSERSTACK_USERNAME'],
            pass: process.env['BROWSERSTACK_ACCESS_KEY']
        },

        qs: (0, _assign2.default)({}, BUILD_ID && { build: BUILD_ID }, PROJECT_NAME && { project: PROJECT_NAME }, params),

        method: apiPath.method || 'GET',
        json: !apiPath.binaryStream
    };

    if (apiPath.binaryStream) opts.encoding = null;

    var currentRequestPromise = apiRequestPromise.then(function () {
        return (0, _requestPromise2.default)(url, opts);
    }).catch(function (error) {
        if (error.statusCode === 401) throw new Error(ERROR_MESSAGES.BROWSERSTACK_AUTHENTICATION_FAILED());

        throw error;
    });

    apiRequestPromise = currentRequestPromise.then(function () {
        return (0, _delay2.default)(API_REQUEST_DELAY);
    });

    return currentRequestPromise;
};

var _pinkie = require('pinkie');

var _pinkie2 = _interopRequireDefault(_pinkie);

var _requestPromise = require('request-promise');

var _requestPromise2 = _interopRequireDefault(_requestPromise);

var _delay = require('./utils/delay');

var _delay2 = _interopRequireDefault(_delay);

var _errorMessages = require('./templates/error-messages');

var ERROR_MESSAGES = _interopRequireWildcard(_errorMessages);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BUILD_ID = process.env['BROWSERSTACK_BUILD_ID'];
var PROJECT_NAME = process.env['BROWSERSTACK_PROJECT_NAME'];

var API_REQUEST_DELAY = 100;

var apiRequestPromise = _pinkie2.default.resolve(null);

module.exports = exports['default'];
